### PR TITLE
Refactoring service construction 

### PR DESCRIPTION
**Pull request by sourabhdalvi**

### CODE REVIEW

File: src/devices_models/devices/AC_branches.jl [hunk 1]
```diff
@@ -39,6 +39,20 @@ get_variable_binary(
 ) = false
 
 get_variable_multiplier(_, ::Type{<:PSY.ACBranch}, _) = NaN
+
+function initialize_timeseries_names(
```
Comment (jd-lara) [inline src/devices_models/devices/AC_branches.jl:43]: I don't understand why we have time series here. No model in branches uses time series.

File: src/core/service_model.jl [hunk 2]
```diff
@@ -85,6 +88,13 @@ get_duals(m::ServiceModel) = m.duals
 get_time_series_names(m::ServiceModel) = m.time_series_names
 get_attributes(m::ServiceModel) = m.attributes
 get_attribute(m::ServiceModel, key::String) = get(m.attributes, key, nothing)
+get_contributing_devices_map(m::ServiceModel) = m.contributing_devices_map
+get_contributing_devices_map(m::ServiceModel, key) =
+    get(m.contributing_devices_map, key, nothing)
+get_contributing_devices(m::ServiceModel) =
```
Comment (jd-lara) [inline src/core/service_model.jl:94]: I don't understand this function. Are you trying to get all the contributing devices?

File: src/devices_models/devices/electric_loads.jl [hunk 3]
```diff
@@ -33,6 +33,23 @@ get_multiplier_value(::TimeSeriesParameter, d::PSY.ElectricLoad, ::AbstractContr
 
 #! format: on
 
+function initialize_timeseries_names(
```
Comment (jd-lara) [inline src/devices_models/devices/electric_loads.jl:36]: @daniel-thom what do you think about these names to set-up the default values of the time series and attributes of the models. Should we re-name this methods to something more clear?

File: src/services_models/services_constructor.jl [hunk 4]
```diff
@@ -41,6 +32,54 @@ function get_incompatible_devices(devices_template::Dict)
...
+    types = unique(typeof.(contributing_devices))
+    device_list = Dict()
+    for t in types
+        _devices = filter(x -> typeof(x) == t, contributing_devices)
+        device_list[t] = _devices
+    end
+    return device_list
+end
+
+function populate_contributing_devices!(
```
Comment (jd-lara) [inline src/services_models/services_constructor.jl:45]: this function might need some additional care. I don't fully understand what it is doing.

File: src/services_models/services_constructor.jl [hunk 5]
```diff
@@ -41,6 +32,54 @@ function get_incompatible_devices(devices_template::Dict)
...
+    container::OptimizationContainer,
```
Comment (jd-lara) [inline src/services_models/services_constructor.jl:46]: where are we using the container here?

File: src/devices_models/devices/electric_loads.jl (hunk 3 above)
Comment (daniel-thom) [reply src/devices_models/devices/electric_loads.jl:36]: I would probably call it `get_time_series_name_definitions`.

Note that we use `time_series` (two words).

Comment (jd-lara) [reply src/devices_models/devices/electric_loads.jl:36]: How about `get_time_series_default_names`

File: src/devices_models/devices/electric_loads.jl [hunk 6]
```diff
@@ -33,6 +33,23 @@ get_multiplier_value(::TimeSeriesParameter, d::PSY.ElectricLoad, ::AbstractContr
...
+    ::Type{<:PSY.ElectricLoad},
+    ::Type{<:Union{FixedOutput, AbstractLoadFormulation}},
+)
+    return Dict{Type{<:TimeSeriesParameter}, String}(
+        ActivePowerTimeSeriesParameter => "max_active_power",
+        ReactivePowerTimeSeriesParameter => "max_active_power",
+    )
+end
+
+function initialize_attributes(
```
Comment (jd-lara) [inline src/devices_models/devices/electric_loads.jl:46]: `get_default_attributes`

File: src/core/service_model.jl (hunk 2 above)
Comment (jd-lara) [reply src/core/service_model.jl:94]: Iterators.Flatten(values(m.contributing_devices_map))

File: src/services_models/services_constructor.jl [hunk 7]
```diff
@@ -41,6 +32,54 @@ function get_incompatible_devices(devices_template::Dict)
     return incompatible_device_types
 end
 
+function sort_contributing_devices(contributing_devices)
```
Comment (jd-lara) [inline src/services_models/services_constructor.jl:35]: `map_by_type_contributing_devices`

File: src/core/service_model.jl (hunk 2 above)
Comment (sourabhdalvi) [reply src/core/service_model.jl:94]: The downstream implementation of add_variables only take a `Vector{PSY.Component}` or  `IS.FlattenIteratorWrapper{PSY.Component}`. So leaving it as is for now

Comment (jd-lara) [reply src/core/service_model.jl:94]: Can you leave a comment in the code please

File: src/devices_models/devices/thermal_generation.jl [hunk 8]
```diff
@@ -28,6 +28,7 @@ get_variable_initial_value(::ActivePowerVariable, d::PSY.ThermalGen, ::AbstractT
 get_variable_initial_value(::PowerAboveMinimumVariable, d::PSY.ThermalGen, ::AbstractCompactUnitCommitment) = max(0.0, PSY.get_active_power(d) - PSY.get_active_power_limits(d).min)
 
 get_variable_lower_bound(::ActivePowerVariable, d::PSY.ThermalGen, ::AbstractThermalFormulation) = PSY.get_active_power_limits(d).min
+get_variable_lower_bound(::ActivePowerVariable, d::PSY.ThermalGen, ::AbstractThermalUnitCommitment) = 0.0
```
Comment (sourabhdalvi) [inline src/devices_models/devices/thermal_generation.jl:31]: So in the previous implementation of range constraints, we would grab the variable and remove its bounds if we added a binary variable based lower bound. Now with the possibility of having either variables or expressions, its better to set the bound to 0.0 here and changing it later

File: src/devices_models/devices/hydro_generation.jl [hunk 9]
```diff
@@ -19,6 +19,7 @@ get_variable_binary(::ActivePowerVariable, ::Type{<:PSY.HydroGen}, ::AbstractHyd
 get_variable_initial_value(::ActivePowerVariable, d::PSY.HydroGen, ::AbstractHydroFormulation) = PSY.get_active_power(d)
 
 get_variable_lower_bound(::ActivePowerVariable, d::PSY.HydroGen, ::AbstractHydroFormulation) = PSY.get_active_power_limits(d).min
+get_variable_lower_bound(::ActivePowerVariable, d::PSY.HydroGen, ::AbstractHydroUnitCommitment) = 0.0
```
Comment (sourabhdalvi) [inline src/devices_models/devices/hydro_generation.jl:22]: same thing as thermal commitment range constraints

Comment (jd-lara) [reply src/devices_models/devices/hydro_generation.jl:22]: ok

File: src/devices_models/devices/thermal_generation.jl (hunk 8 above)
Comment (jd-lara) [reply src/devices_models/devices/thermal_generation.jl:31]: ok

**Comment by jd-lara** [comment]
Just tests locally and all tests passed

**Pull request merged**